src/sentry/api/endpoints/auth_index.py

```diff
@@ -35,6 +35,85 @@ class AuthIndexEndpoint(Endpoint):
 
     permission_classes = ()
 
+    @staticmethod
+    def _reauthenticate_with_sso(request, org_id):
+        """
+        If a user without a password is hitting this, it means they need to re-identify with SSO.
+        """
+        redirect = request.META.get("HTTP_REFERER", None)
+        if not is_safe_url(redirect, allowed_hosts=(request.get_host(),)):
+            redirect = None
+        initiate_login(request, redirect)
+        raise SsoRequired(Organization.objects.get_from_cache(id=org_id))
+
+    @staticmethod
+    def _verify_user_via_inputs(validator, request):
+        # See if we have a u2f challenge/response
+        if "challenge" in validator.validated_data and "response" in validator.validated_data:
+            try:
+                interface = Authenticator.objects.get_interface(request.user, "u2f")
+                if not interface.is_enrolled():
+                    raise LookupError()
+                challenge = json.loads(validator.validated_data["challenge"])
+                response = json.loads(validator.validated_data["response"])
+                authenticated = interface.validate_response(request, challenge, response)
+                if not authenticated:
+                    logger.warning(
+                        "u2f_authentication.verification_failed",
+                        extra={"user": request.user.id},
+                    )
+                return authenticated
+            except ValueError as err:
+                logger.warning(
+                    "u2f_authentication.value_error",
+                    extra={"user": request.user.id, "error_message": err},
+                )
+                pass
+            except LookupError:
+                logger.warning(
+                    "u2f_authentication.interface_not_enrolled",
+                    extra={"validated_data": validator.validated_data, "user": request.user.id},
+                )
+                pass
+        # attempt password authentication
+        elif "password" in validator.validated_data:
+            authenticated = request.user.check_password(validator.validated_data["password"])
+            return authenticated
+        return False
+
+    def _validate_superuser(self, validator, request):
+        """
+        For a superuser, they need to be validated before we can grant an active superuser session.
+        If the user has a password or u2f device, authenticate the password/challenge that was sent is valid.
+        If the user doesn't have a password or u2f device, we say they're authenticated if they have a
+        valid SSO session.
+
+        By nature of granting an active superuser session, we want to make sure that the user has completed
+        SSO and if they do not, we redirect them back to the SSO login.
+
+        """
+        # TODO Look at AuthVerifyValidator
```

Comment:
Does this need attention before shipping?

Reply:
nope!